src/substantial/src/converters.rs

```diff
@@ -125,6 +239,97 @@ impl Run {
     pub fn reset(&mut self) {
         self.operations = vec![];
     }
+
+    pub fn next_id(&mut self) -> u32 {
+        self.id += 1;
+        self.id
+    }
+
+    pub fn append_op(&mut self, op: OperationEvent) {
+        let has_stopped = self
+            .operations
+            .iter()
+            .any(|op| matches!(op.event, OperationEvent::Stop { .. }));
+        if !has_stopped {
+            self.operations.push(Operation {
+                at: Utc::now(),
+                event: op,
+            });
+        }
+    }
+
+    pub fn save(&mut self) -> Result<SaveOutput> {
+        let next_id = self.next_id();
+        let mut current_retry_count: i32 = 1;
+
+        for Operation { event, .. } in self.operations.iter() {
+            if let OperationEvent::Save { id, value } = event {
+                if *id == next_id {
+                    if let SavedValue::Resolved { payload } = value {
+                        return Ok(SaveOutput {
+                            payload: Some(payload.clone()),
+                            current_retry_count: None,
+                        });
+                    } else if let SavedValue::Retry {
+                        counter,
+                        wait_until,
+                    } = value
+                    {
+                        let now = Utc::now();
+                        if wait_until > &now {
+                            bail!(Interupt::Saveretry);
```

Comment:
I suggest you to use enums directly. As we are within rust already, we can make clear distinctions between execution errors, interrupts and the returned value.